contracts/VToken.sol

```diff
@@ -1109,61 +1109,109 @@ contract VToken is Ownable2StepUpgradeable, VTokenInterface, ExponentialNoError,
+        accountStableBorrows[account].stableRateMantissa = stableRateMantissaNew;
 
-            accountBorrows[account].principal = 0;
-            accountBorrows[account].interestIndex = borrowIndex;
-        } else {
-            require(stableDebt > 0, "vToken: swapBorrowRateMode stable debt is 0");
+        accountBorrows[account].principal = variableDebt - swappedAmount;
+        accountBorrows[account].interestIndex = borrowIndex;
 
-            stableBorrowsNew = stableBorrows - stableDebt;
+        return (stableBorrowsNew, averageStableBorrowRateNew);
+    }
 
-            uint256 stableRateMantissa = accountStableBorrows[account].stableRateMantissa;
+    function _updateStatesForVariableRateSwap(
+        uint256 swappedAmount,
+        uint256 stableDebt,
+        uint256 variableDebt,
+        address account
+    ) internal returns (uint256, uint256) {
+        uint256 newStableDebt = stableDebt - swappedAmount;
+        uint256 stableBorrowsNew = stableBorrows - swappedAmount;
 
-            if (stableBorrowsNew == 0) {
-                averageStableBorrowRateNew = 0;
-            } else {
-                unchecked {
-                    averageStableBorrowRateNew =
-                        ((stableBorrows * averageStableBorrowRate) - (stableDebt * stableRateMantissa)) /
-                        stableBorrowsNew;
-                }
+        uint256 stableRateMantissa = accountStableBorrows[account].stableRateMantissa;
+        uint256 averageStableBorrowRateNew;
+        if (stableBorrowsNew == 0) {
+            averageStableBorrowRateNew = 0;
+        } else {
+            unchecked {
+                averageStableBorrowRateNew =
+                    ((stableBorrows * averageStableBorrowRate) - (swappedAmount * stableRateMantissa)) /
+                    stableBorrowsNew;
             }
+        }
 
-            /////////////////////////
-            // EFFECTS & INTERACTIONS
-            // (No safe failures beyond this point)
-            accountBorrows[account].principal = accountBorrowsNew;
-            accountBorrows[account].interestIndex = borrowIndex;
+        /////////////////////////
+        // EFFECTS & INTERACTIONS
+        // (No safe failures beyond this point)
+        accountBorrows[account].principal = variableDebt + swappedAmount;
+        accountBorrows[account].interestIndex = borrowIndex;
 
-            accountStableBorrows[account].principal = 0;
-            accountStableBorrows[account].interestIndex = stableBorrowIndex;
+        accountStableBorrows[account].principal = newStableDebt;
+        accountStableBorrows[account].interestIndex = stableBorrowIndex;
+
+        return (stableBorrowsNew, averageStableBorrowRateNew);
+    }
+
+    /**
+     * @dev Allows a borrower to swap his debt between stable and variable mode, or vice versa
+     * @param rateMode The rate mode that the user wants to swap to
+     **/
+    function swapBorrowRateMode(uint256 rateMode) external {
```

Comment:
We could try to reuse `swapBorrowRateModeWithAmount`, calling it from this function with `type(uint256).max` (and supporting that in the called function), for example.

Reply:
`swapBorrowRateModeWithAmount` have addition two checks that's why I created seperate method `swapBorrowRateMode`

Reply:
The added check is `require(variableDebt >= amount, "Insufficient amount");`, that would be satisfied if we properly support `type(uint256).max`.

Even more, I was thinking we could remove this function from the contract, delegating in the client the option to invoke `swapBorrowRateModeWithAmount` (I would rename it) with `type(uint256).max` if they want to swap 100% of the debt, as we are doing with the repay function, for example.